File: src/main/java/org/veupathdb/lib/blast/field/TemplateLength.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.JSONConstructor;
import org.veupathdb.lib.blast.util.JSONValue;

public enum TemplateLength implements JSONValue
{
  Length16(16),
  Length18(18),
  Length21(21);

  private final int value;

  TemplateLength(int value) {
    this.value = value;
  }

  public int getValue() {
    return value;
  }

  @Override
  public JsonNode toJSON() {
    return JSONConstructor.newInt(value);
  }

  public static TemplateLength fromIntValue(int value) {
    for (var val : values())
      if (val.getValue() == value)
        return val;

    throw new IllegalArgumentException(
      "Invalid template_length value \"" + value + "\", must be one of 16, 18, or 21."
    );
  }

  public static TemplateLength fromJSON(JsonNode js) {
    return fromIntValue(js.intValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/TemplateLength.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.JSONConstructor.newInt
import org.veupathdb.lib.blast.util.JSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.annotation.JsonCreator
import java.lang.IllegalArgumentException

enum class TemplateLength(val value: Int) : JSONValue {
  Length16(16), Length18(18), Length21(21);

  @JsonValue
  override fun toJSON(): JsonNode = newInt(value)

  companion object {
    @JvmStatic
    fun fromIntValue(value: Int): TemplateLength {
      for (v in values())
        if (v.value == value)
          return v

      throw IllegalArgumentException(
        "Invalid template_length value \"$value\", must be one of 16, 18, or 21."
      )
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): TemplateLength {
      return fromIntValue(js.intValue())
    }
  }
}